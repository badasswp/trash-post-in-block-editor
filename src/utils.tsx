import apiFetch from '@wordpress/api-fetch';
import { select, dispatch } from '@wordpress/data';

/**
 * Trash Post
 *
 * This function trashes the post based
 * on the Post ID.
 *
 * @since 1.0.0
 *
 * @return {void}
 */
export const trashPost = async () => {
	const { getCurrentPostId } = select( 'core/editor' );
	const { createWarningNotice } = dispatch( 'core/notices' ) as any;

	try {
		await apiFetch( {
			path: '/tpbe/v1/trash',
			method: 'POST',
			data: {
				id: getCurrentPostId(),
			},
		} );

		window.location.href = `${ tpbe.url }`;
	} catch ( e ) {
		createWarningNotice( e );
	}
};
