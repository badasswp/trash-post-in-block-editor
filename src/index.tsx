import { __ } from '@wordpress/i18n';
import { Fragment, useState } from '@wordpress/element';
import { registerPlugin } from '@wordpress/plugins';
import { PluginSidebar } from '@wordpress/edit-post';
import { Modal, PanelBody, Button } from '@wordpress/components';

import { trashPost } from './utils';

/**
 * Trash Post In Block Editor.
 *
 * This function returns a JSX component that comprises
 * the Plugin Sidebar and Trash icon.
 *
 * @since 1.0.0
 *
 * @returns {JSX.Element}
 */
const TrashPostInBlockEditor = () => {
  const { getCurrentPostId } = select('core/editor');

  const trashPost = async () => {
    const postID = getCurrentPostId();

    try {
      await apiFetch(
        {
          path: '/tpbe/v1/trash',
          method: 'POST',
          data: {
            id: postID
          },
        }
      );

      window.location.href = `${tpbe.url}`
    } catch (e) {
      console.log(e);
    }
  }

  return (
    <Fragment>
      <PluginSidebar
        name="tpbe-sidebar"
        title={ __( 'Trash Post in Block Editor', 'trash-post-in-block-editor' ) }
        icon="trash"
      >
        <PanelBody>
          <div id="tpbe">
            <Button
              variant="primary"
              onClick={trashPost}
            >
              { __( 'Trash Post', 'trash-post-in-block-editor' ) }
            </Button>
          </div>
        </PanelBody>
      </PluginSidebar>
    </Fragment>
  );
};

registerPlugin( 'trash-post-in-block-editor', {
  render: TrashPostInBlockEditor,
} );
