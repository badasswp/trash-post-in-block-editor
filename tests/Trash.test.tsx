import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import TrashPostInBlockEditor from '../src/index';
import { trashPost } from '../src/utils';

jest.mock( '../src/utils', () => ( {
	trashPost: jest.fn(),
} ) );

describe( 'TrashPostInBlockEditor', () => {
	beforeEach( () => {
		global.tpbe = {
			wpVersion: '6.6',
			url: 'https://example.com/wp-admin/edit.php',
		};
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'matches snapshot', () => {
		const { container } = render( <TrashPostInBlockEditor /> );

		expect( container ).toMatchSnapshot();
	} );

	it( 'renders trash button in pinned item area', () => {
		const { getByTestId } = render( <TrashPostInBlockEditor /> );

		// Test if Trash Post button is displayed.
		const trashButton = getByTestId( 'tpbe-trash-btn' );
		expect( trashButton ).toBeVisible();
	} );

	it( 'displays modal when trash button is clicked', () => {
		const { getByTestId, getByText, getByRole } = render(
			<TrashPostInBlockEditor />
		);

		// Click Trash Post button to open modal.
		fireEvent.click( getByTestId( 'tpbe-trash-btn' ) );

		// Test that modal content is displayed.
		expect(
			getByText( 'Are you sure you want to delete this Post?' )
		).toBeVisible();
		expect( getByRole( 'button', { name: 'Yes' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'No' } ) ).toBeVisible();
	} );

	it( 'calls trashPost function if the `Yes` button is clicked', () => {
		const { getByTestId, getByText, getByRole } = render(
			<TrashPostInBlockEditor />
		);

		// Click Trash Post button to open modal.
		fireEvent.click( getByTestId( 'tpbe-trash-btn' ) );

		// Test that modal content is displayed.
		expect(
			getByText( 'Are you sure you want to delete this Post?' )
		).toBeVisible();
		expect( getByRole( 'button', { name: 'Yes' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'No' } ) ).toBeVisible();

		// click Yes button.
		fireEvent.click( getByRole( 'button', { name: 'Yes' } ) );

		// Test that trashPost function is called.
		expect( trashPost ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'closes modal if the `No` button is clicked', () => {
		const { getByTestId, getByText, getByRole, queryByText } = render(
			<TrashPostInBlockEditor />
		);

		// Click Trash Post button to open modal.
		fireEvent.click( getByTestId( 'tpbe-trash-btn' ) );

		// Test that modal content is displayed.
		expect(
			getByText( 'Are you sure you want to delete this Post?' )
		).toBeVisible();
		expect( getByRole( 'button', { name: 'Yes' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'No' } ) ).toBeVisible();

		// Click No button.
		fireEvent.click( getByRole( 'button', { name: 'No' } ) );

		// Test that modal content is no longer in the document.
		expect(
			queryByText( 'Are you sure you want to delete this Post?' )
		).not.toBeInTheDocument();
	} );
} );
