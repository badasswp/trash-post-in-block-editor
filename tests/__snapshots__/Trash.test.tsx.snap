// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrashPostInBlockEditor renders sidebar and Trash Post button 1`] = `
<div>
  <div
    class="editor-sidebar"
  >
    <h2
      class="interface-complementary-area-header__title"
    >
      Trash Post in Block Editor
    </h2>
    <div
      class="components-panel__body is-opened"
    >
      <div
        id="tpbe"
      >
        <button
          class="components-button is-primary"
          type="button"
        >
          Trash Post
        </button>
      </div>
    </div>
  </div>
</div>
`;
